### ISSUE TITLE
Understanding Encoder And Decoder LLMs

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#10557]
Understanding Encoder And Decoder LLMs<br>
<br>
https://ift.tt/xE7DIdQ<br>
<br>
Sebastian Raschka, PhD<br>
<br>

<div><div><div><div><div dir="auto">
<p>Several people asked me to dive a bit deeper into large language model (LLM) jargon and explain some of the more technical terms we nowadays take for granted. This includes references to "encoder-style" and "decoder-style" LLMs. What do these terms mean? </p>
<p><span>To explain the difference between encoder- and decoder-style LLMs, I wanted to share an excerpt from my new book, </span><a href="https://leanpub.com/machine-learning-q-and-ai">Machine Learning Q and AI</a><span>, that I completed last week. </span></p>
<p></p>
<p></p>
<p>This book is aimed at people who are already familiar with machine learning and deep learning ("AI") and are interested in diving into more advanced topics. There are 30 chapters in total, covering various topics, including</p>
<p><span>Right now, the book is only available as a </span><strong><a href="https://leanpub.com/machine-learning-q-and-ai">digital version on Leanpub</a></strong><span>, but a high-quality print version via No Starch Press will follow later this summer!</span></p>
<p><span>Let's get to it: </span><em><strong>What are the differences between encoder- and decoder-based language transformers?</strong></em></p>
<p></p>
<p></p>
<hr>
<p></p>
<p>Fundamentally, both encoder- and decoder-style architectures use the same self-attention layers to encode word tokens. However, the main difference is that encoders are designed to learn embeddings that can be used for various predictive modeling tasks such as classification. In contrast, decoders are designed to generate new texts, for example, answering user queries.</p>
<p><strong>The original transformer</strong></p>
<p><span>The original transformer architecture (</span><em><a href="https://arxiv.org/abs/1706.03762">Attention Is All You Need, 2017</a></em><span>), which was developed for English-to-French and English-to-German language translation, utilized both an encoder and a decoder, as illustrated in the figure below.</span></p>
<p>In the figure above, the input text (that is, the sentences of the text that is to be translated) is first tokenized into individual word tokens, which are then encoded via an embedding layer before it enters the encoder part. Then, after adding a positional encoding vector to each embedded word, the embeddings go through a multi-head self-attention layer. The multi-head attention layer is followed by an "Add &amp; normalize" step, which performs a layer normalization and adds the original embeddings via a skip connection (also known as a residual or shortcut connection). Finally, after entering a "fully connected layer," which is a small multilayer perceptron consisting of two fully connected layers with a nonlinear activation function in between, the outputs are again added and normalized before they are passed to a multi-head self-attention layer of the decoder part.</p>
<p>The decoder part in the figure above has a similar overall structure as the encoder part. The key difference is that the inputs and outputs are different. The encoder receives the input text that is to be translated, and the decoder generates the translated text.</p>
<p><strong>Encoders</strong></p>
<p>The encoder part in the original transformer, illustrated in the preceding figure, is responsible for understanding and extracting the relevant information from the input text. It then outputs a continuous representation (embedding) of the input text that is passed to the decoder. Finally, the decoder generates the translated text (target language) based on the continuous representation received from the encoder.</p>
<p><span>Over the years, various encoder-only architectures have been developed based on the encoder module of the original transformer model outlined above. Notable examples include BERT (</span><em><a href="https://arxiv.org/abs/1810.04805">Pre-training of Deep Bidirectional Transformers for Language Understanding, 2018</a></em><span>) and RoBERTa (</span><em><a href="https://arxiv.org/abs/1907.11692">A Robustly Optimized BERT Pretraining Approach, 2018</a></em><span>).</span></p>
<p></p>
<p><span>BERT (</span><strong>B</strong><span>idirectional </span><strong>E</strong><span>ncoder </span><strong>R</strong><span>epresentations from </span><strong>T</strong><span>ransformers) is an encoder-only architecture based on the Transformer's encoder module. The BERT model is pretrained on a large text corpus using masked language modeling (illustrated in the figure below) and next-sentence prediction tasks.</span></p>
<div><figure><a href="https://substackcdn.com/image/fetch/f_auto,q_auto:good,fl_progressive:steep/https%3A%2F%2Fsubstack-post-media.s3.amazonaws.com%2Fpublic%2Fimages%2Fb8fc5971-d59b-4d5a-bb2a-d8ac6d2aa23a_2022x1224.png" target="_blank"></a><figcaption>Illustration of the masked language modeling pretraining objective used in BERT-style transformers.</figcaption></figure></div>
<p>The main idea behind masked language modeling is to mask (or replace) random word tokens in the input sequence and then train the model to predict the original masked tokens based on the surrounding context.</p>
<p>Next to the masked language modeling pretraining task illustrated in the figure above, the next-sentence prediction task asks the model to predict whether the original document's sentence order of two randomly shuffled sentences is correct. For example, two sentences, in random order, are separated by the [SEP] token:</p>
<ul>
<li><p>[CLS] Toast is a simple yet delicious food [SEP] It’s often served with butter, jam, or honey.</p></li>
<li><p>[CLS] It’s often served with butter, jam, or honey. [SEP] Toast is a simple yet delicious food.</p></li>
</ul>
<p><span>The [CLS] token is a placeholder token for the model, prompting the model to return a </span><em>True</em><span> or </span><em>False</em><span> label indicating whether the sentences are in the correct order or not.</span></p>
<p>The masked language and next-sentence pretraining objectives (which are a form of self-supervised learning, as discussed in Chapter 2) allow BERT to learn rich contextual representations of the input texts, which can then be finetuned for various downstream tasks like sentiment analysis, question-answering, and named entity recognition.</p>
<p><span>RoBERTa (</span><strong>R</strong><span>obustly </span><strong>o</strong><span>ptimized </span><strong>BERT</strong><span> </span><strong>a</strong><span>pproach) is an optimized version of BERT. It maintains the same overall architecture as BERT but employs several training and optimization improvements, such as larger batch sizes, more training data, and eliminating the next-sentence prediction task. These changes resulted in RoBERTa achieving better performance on various natural language understanding tasks than BERT.</span></p>
<p><strong>Decoders</strong></p>
<p><span>Coming back to the original transformer architecture outlined at the beginning of this section, the multi-head self-attention mechanism in the decoder is similar to the one in the encoder, but it is masked to prevent the model from attending to future positions, ensuring that the predictions for position </span><em>i</em><span> can depend only on the known outputs at positions less than </span><em>i</em><span>. As illustrated in the figure below, the decoder generates the output word by word.</span></p>
<div><figure><a href="https://substackcdn.com/image/fetch/f_auto,q_auto:good,fl_progressive:steep/https%3A%2F%2Fsubstack-post-media.s3.amazonaws.com%2Fpublic%2Fimages%2Fd3c55ad5-cc86-4421-b0bc-bbd967409ca6_1162x1022.png" target="_blank"></a><figcaption>Illustration of the next-sentence prediction task used in the original transformer.</figcaption></figure></div>
<p>This masking (shown explicitly in the figure above, although it happens internally in the decoder's multi-head self-attention mechanism) is essential to maintain the autoregressive property of the transformer model during training and inference. The autoregressive property ensures that the model generates output tokens one at a time and uses previously generated tokens as context for generating the next word token.</p>
<p>Over the years, researchers have built upon the original encoder-decoder transformer architecture and developed several decoder-only models that have proven to be highly effective in various natural language processing tasks. The most notable models include the GPT family.</p>
<p><span>The GPT (</span><strong>G</strong><span>enerative </span><strong>P</strong><span>re-trained </span><strong>T</strong><span>ransformer) series are decoder-only models pretrained on large-scale unsupervised text data and finetuned for specific tasks such as text classification, sentiment analysis, question-answering, and summarization. The GPT models, including GPT-2, (</span><em><a href="https://arxiv.org/abs/2005.14165">GPT-3 Language Models are Few-Shot Learners, 2020</a></em><span>), and the more recent GPT-4, have shown remarkable performance in various benchmarks and are currently the most popular architecture for natural language processing.</span></p>
<p>One of the most notable aspects of GPT models is their emergent properties. Emergent properties refer to the abilities and skills that a model develops due to its next-word prediction pretraining. Even though these models were only taught to predict the next word, the pretrained models are capable of text summarization, translation, question answering, classification, and more. Furthermore, these models can perform new tasks without updating the model parameters via in-context learning, which is discussed in more detail in Chapter 18.</p>
<p><strong>Encoder-decoder hybrids</strong></p>
<p>Next to the traditional encoder and decoder architectures, there have been advancements in the development of new encoder-decoder models that leverage the strengths of both components. These models often incorporate novel techniques, pre-training objectives, or architectural modifications to enhance their performance in various natural language processing tasks. Some notable examples of these new encoder-decoder models include </p>
<p>Encoder-decoder models are typically used for natural language processing tasks that involve understanding input sequences and generating output sequences, often with different lengths and structures. They are particularly good at tasks where there is a complex mapping between the input and output sequences and where it is crucial to capture the relationships between the elements in both sequences. Some common use cases for encoder-decoder models include text translation and summarization.</p>
<p><strong>Terminology and jargon</strong></p>
<p><span>All of these methods, encoder-only, decoder-only, and encoder-decoder models, are sequence-to-sequence models (often abbreviated as </span><em>seq2seq</em><span>). Note that while we refer to BERT-style methods as encoder-only, the description </span><em>encoder-only</em><span> may be misleading since these methods also </span><em>decode</em><span> the embeddings into output tokens or text during pretraining.</span></p>
<p>In other words, both encoder-only and decoder-only architectures are "decoding." However, the encoder-only architectures, in contrast to decoder-only and encoder-decoder architectures, are not decoding in an autoregressive fashion. Autoregressive decoding refers to generating output sequences one token at a time, conditioning each token on the previously generated tokens. Encoder-only models do not generate coherent output sequences in this manner. Instead, they focus on understanding the input text and producing task-specific outputs, such as labels or token predictions.</p>
<p></p>
<p><strong>Conclusion</strong></p>
<p>In brief, encoder-style models are popular for learning embeddings used in classification tasks, encoder-decoder-style models are used in generative tasks where the output heavily relies on the input (for example, translation and summarization), and decoder-only models are used for other types of generative tasks including Q&amp;A. Since the first transformer architecture emerged, hundreds of encoder-only, decoder-only, and encoder-decoder hybrids have been developed, as summarized in the figure below.</p>
<div><figure><a href="https://substackcdn.com/image/fetch/f_auto,q_auto:good,fl_progressive:steep/https%3A%2F%2Fsubstack-post-media.s3.amazonaws.com%2Fpublic%2Fimages%2Fcce3c437-4b9c-4d15-947d-7c177c9518e5_4258x5745.png" target="_blank"></a><figcaption><span>An overview of </span><em><strong>some</strong></em><span> of the most popular large language transformers organized by architecture type and developers.</span></figcaption></figure></div>
<p></p>
<p>While encoder-only models gradually lost in popularity, decoder-only models like GPT exploded in popularity thanks to breakthrough in text generation via GPT-3, ChatGPT, and GPT-4. However, encoder-only models are still very useful for training predictive models based on text embeddings versus generating texts.</p>
<p></p>
<p></p>
<hr>
<p><span>I hope you found this article useful! If you’d like to read more and support my work, please consider buying a copy of my book </span><em><strong><a href="https://leanpub.com/machine-learning-q-and-ai">Machine Learning and AI</a></strong></em><span> or consider a paid subscription on Substack.</span></p>
<p></p>
<p>PS: As a thank you to the Ahead of AI supporters, I included a coupon code for $10 off until the end of the month below.</p>
<p></p>
</div></div></div></div></div><br>
<br>
<br>
<br>
<br>
<br>
via  <br>
<br>
January 15, 2024 at 09:33AM
